### PR TITLE
Scripting support

### PR DESCRIPTION
**Pull request by IMcD23**
Major changes:
- Moved command execution to a new CommandExecutor class, which calls into ios_system, and can run multiple commands at once (in series, put on a dispatch queue).
- Streaming stdout/stderr to UITextView as it comes in. Using a Pipe & File Handle to accomplish this.
- Added ScriptsViewController & ScriptEditViewController, for managing and editing scripts in a GUI.
- Added Script class, that handles script management (creating/loading/editing/running). Scripts are stored in Documents/.scripts
- Refactored AutoCompleteManagerDelegate & InputAssistantViewDataSource conformances into a shared file.
- Refactored TerminalView's UITextView into a UITextView subclass "TerminalTextView", which is shared between terminal view and ScriptEditViewController.
- Removed TerminalProcessor protocol (replaced by CommandExecutor)

Validation:
Tested basic commands, running scripts on iPhone X (iOS 11.3), iPad Pro 9.7" (iOS 11.2.1)
No regressions found. All gui for scripts work.

### CODE REVIEW

**Comment by IMcD23** [comment]
Additional testing will be necessary, especially with the stdout/stderr changes also made in this PR. 

@palmin, I see you have [another PR](https://github.com/louisdh/terminal/pull/52) open for similar changes. This will probably conflict / need validation with yours.

@louisdh Let me know if you have any questions on how this stuff works. Thanks!

**Comment by palmin** [comment]
@IMcD23 My pull request was just a one-line fix to the `open-url` command make sure errors appeared on thread_stderr and not the generic stderr. 

It does nothing outside this `open-url` command and since your changes have not touched this file, I think the changes will be compatible.

**Comment by IMcD23** [comment]
Added a `CommandExecutorCommand` protocol, similar to the class from #40, but as a protocol as suggested. There are currently 3 executor commands conforming to that protocol: empty (no-op), system (run in ios_system), and script (run a script). These commands can call each other, which allows for scripts running other scripts and various commands.

**Comment by louisdh** [comment]
Awesome stuff! This is a great start for scripting, I will write out my thoughts for more advanced scripting at issue #48.

**Comment by holzschu** [comment]
I might need help here. CommandExecutor breaks one assumption in `iOS_system`, namely that if `thread_stdout == stdout`, then we are in the main window, otherwise we are in a pipe or with redirection. It's the equivalent of `isatty()`, which of course doesn't work.
Many commands adapt their output depending on this, `ls` being the most evident: multi-column output if in a window, single-column output if redirected (so `ls | grep \.cpp` is useful). 

Short version: I need a way to know, inside `ios_system`, whether this command outputs to the main window or to a program or a file.

**Comment by IMcD23** [comment]
Ahh makes sense. I saw that check, and assumed that it would handle it properly, not realizing that different outputs are for pipes.

Perhaps expose `window_stdout`, `window_stderr`, `window_stdin`, that default to `stdout`, `stderr`, `stdin` respectively?

**Comment by holzschu** [comment]
Simply setting `stdout` to `executor.stdout_file` before calling `ios_system` restores the previous behaviour. Now I need to make sure I only do it for the main command, not when it's in a script.

**Comment by holzschu** [comment]
Am I correct in assuming that `dispatch(command)` is only called from `ViewController` (main window) and that scripts actually call `run(command)`?

**Comment by IMcD23** [comment]
Correct. `dispatch` is called when the user hits the enter key. `run` is what is called on the execution thread, which does the actual work.
